### DISCUSSION TITLE
Order by column from child relation produces not valid SQL request

### DISCUSSION THREAD

**Discussion by Ujifman** [sqlalchemy/sqlalchemy#11906] [Usage Questions]
I have 2 entities, Parent and Children, one to one mapping. I need to get Parents list with filtering and ordering by both parent and children fields.

Here is models:

```python
parent_table = Table(
    'parent',
    mapper_registry.metadata,
    Column('id', Integer, primary_key=True, autoincrement=True, index=True),
    Column('name', UUID(as_uuid=True), nullable=False, unique=True, index=True),
)

children_table = Table(
    'children',
    mapper_registry.metadata,
    Column('id', Integer, primary_key=True, autoincrement=True, index=True),
    Column(
        'parent_id',
        ForeignKey('parent.id'),
        nullable=False,
        index=True,
    ),
    Column('some_name', String, nullable=True, index=True),
)

class Children(Base):
    __table__ = children_table

class Parent(Base):
    __table__ = parent_table

    children: Mapped[Children] = relationship(
        Children,
        uselist=False,
        cascade='all, delete',
        lazy='joined',
    )
```

So when i call filtering, a get valid SQL request, and ORM add children join for selecting fields, end second join for filtering fields, but when I call order by children fields, I get invalid SQL, because ORM add only one join for selecting fields, but not for ordering.
Examples of SQL requests

- filter python call `query = select(Parent).filter(Children.some_name == 'DUMMY').limit(20).offset(0)`

```sql
SELECT
    parent.id,
    parent.name,
    children_1.id AS id_1,
    children_1.parent_id,
    children_1.some_name
FROM
    parent
    LEFT OUTER JOIN children AS children_1 ON parent.id = children_1.parent_id,
    children
WHERE
    children.some_name = 'DUMMY'
LIMIT
    20 OFFSET 0
```

- order by python call `query = select(Parent).order_by(Children.some_name).limit(20).offset(0)`

```sql
SELECT
    parent.id,
    parent.name,
    children_1.id AS id_1,
    children_1.parent_id,
    children_1.some_name
FROM
    parent
    LEFT OUTER JOIN children AS children_1 ON parent.id = children_1.parent_id,
ORDER BY
    children.some_name
LIMIT
    20 OFFSET 0
```

In second query missed second join of `children`, and ORDER BY `children.some_name` refers to undefined table alias, so I get `ProgrammingError('(psycopg2.errors.UndefinedTable) invalid reference to FROM-clause entry ... Perhaps you meant to reference the table alias "extra_fields_1"`.

I can fix it by adding explicit join, but it is bad decision for code maintainability in case of increasing child classes amount. I suppose that order by should work same as with filter query, and building broken SQL request is a bug.

**Comment by CaselIT** [comment]
Hi,

both of these query seems to be equally wrong, since in the first one it's doing a cross join.

In both cases if you want to use child tables in the query join with it: ` select(Parent).join(Parent.children)...`

> but it is bad decision for code maintainability in case of increasing child classes amount

I disagree. You want a join so the query you do should have a join() call

**Reply by Ujifman** [reply]
> Hi,
> 
> both of these query seems to be equally wrong, since in the first one it's doing a cross join.
> 
> In both cases if you want to use child tables in the query join with it: ` select(Parent).join(Parent.children)...`
> 
> > but it is bad decision for code maintainability in case of increasing child classes amount
> 
> I disagree. You want a join so the query you do should have a join() call

But if I select without any filters, than SqlAlchemy itself makes join, because of `lazy='joined',` parameter, and I can't affect it, so if I do explicit join, than in SQL there will be 2 joins, that is not good from the point of performance, because db engine makes +1 unnecessary join.
Anyway I expected, that filter and orderby would use children_1, instead unnecessary +1 join

**Reply by CaselIT** [reply]
Please see https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#the-zen-of-joined-eager-loading for the reason why it doesn't work that way

**Reply by CaselIT** [reply]
note that you can use `contains_eager` in that case to avoid the second join that comes from the lazy='join'
